lib/util/ast.js

```diff
@@ -37,6 +68,49 @@ function findReturnStatement(node) {
   }(bodyNodes));
 }
 
+/**
+ * Helper function for traversing "returns" (return statements or the
+ * returned expression in the case of an arrow function) of a function
+ *
+ * @param {ASTNode} ASTnode The AST node being checked
+ * @param {function} enterFunc Function to execute for each returnStatement found
+ */
+function traverseReturns(ASTnode, enterFunc) {
+  switch (ASTnode.type) {
+    case 'FunctionExpression':
+    case 'FunctionDeclaration':
+    case 'MethodDefinition':
+      break;
+    case 'ReturnStatement':
+      enterFunc(ASTnode);
+      return;
+    case 'ArrowFunctionExpression':
+      if (ASTnode.expression) {
+        enterFunc(ASTnode.body);
+        return;
+      }
+      break;
+    default:
+      // We expect only to receive function nodes
+      return;
+  }
```

Comment:
```suggestion
  if (ASTNode.type === 'ReturnStatement') {
    enterFunc(ASTnode);
  } else if (ASTNode.type === 'ArrowFunctionExpression') {
    if (ASTNode.expression) {
      enterFunc(ASTnode.body);
    }
  } else if (ASTNode.type === 'ReturnStatement') {
    enterFunc(ASTnode);
  } else if (ASTNode.type !== 'FunctionExpression' && ASTNode.type !== 'FunctionDeclaration' && ASTNode.type !== 'MethodDefinition') {
    throw new TypeError('only function nodes are expected');
  }
```

Reply:
I refactored this with "If-return"s. I think it's a bit cleaner than a chain of 'else if".

`isReturningJSX` (lib/util/Components.js) sometimes is called with `ClassDeclaration` nodes. Now only a few `no-typos` tests were broken by this (I fixed it). IMHO, doing nothing here is safer than throwing an exception.